signal_op/extract_events.m

```diff
@@ -0,0 +1,174 @@
+event_times = idx_event_frames_v3;
+
+
+%%%%%%%%%%%%%%
+%visualization
+%%%%%%%%%%%%%%
+if ~isempty(varargin)
+    if varargin{1}==1
+        figure,
+        plot(h_1stround,'k');
+        hold on
+        plot(h_2ndround,'r');
+        hold off
+        legend('the characteristic event, 1st iteration','the characteristic event, 2nd iteration');
+        xlabel('frame')
+
+        trace = trace / max(trace);
+        mad_scale_coarse = compute_mad(trace);
+        threshold_coarse = coarse_thresh_multiplier*mad_scale_coarse;
+
+        filtered =  conv(trace,g_2ndround,'same');
+        filtered = filtered / max(filtered);
+        mad_scale_fine = compute_mad(filtered);
+        threshold_fine = fine_thresh_multiplier*mad_scale_fine;
+
+        spikes_vec_fine = zeros(1,length(trace));
+        spikes_vec_fine(idx_event_frames_v3) = 1;
+        spikes_vec_coarse = zeros(1,length(trace));
+        spikes_vec_coarse(idx_event_frames) = 1;
+
+        figure,
+        plot(trace,'k');
+        hold on
+        plot(filtered,'m');
+        plot(1:length(trace),repmat(threshold_coarse,1,length(trace)),'--b');%coarse threshold
+        plot(1:length(trace),repmat(threshold_fine,1,length(trace)),'--r');%fine threshold
+        stem(spikes_vec_fine*threshold_fine);
+        stem(spikes_vec_coarse*threshold_coarse);
+        hold off
+        xlabel('frame')
+        ylabel('a.u.')
+        title('raw and deconvolved ica traces together with the thresholds and the detected events')
+        legend('raw trace','deconvolved trace','coarse threshold','deconv threshold')
+    end
+end
+
+%%%%%%%%%%%%%%%%%%%
+%internal functions
+%%%%%%%%%%%%%%%%%%%
+function out = iterate_event_indices(idx_event_frames_in)
+
+    h_len = num_frames_before_max+num_frames_after_max+1;
+    %surrogate waveform
+    event_mat = construct_events(idx_event_frames_in);
+    %calculate the mean of all the trials in the event_mat discarding zeros
+    h = zeros(h_len,1);
+    for k = 1:h_len
+        dum = event_mat(:,k);
+        nonzeros = sum(dum~=0);
+        if nonzeros>=1
+            h(k) = sum(dum)/nonzeros;
+        end
+    end
+    %optional: smooth h before inverting
+    SGfilter = SGsmoothingfilter(20,8);
+    h = conv(h,SGfilter,'same');
+    g_len= h_len; %try to find an inverse g that has the same length as h
+    g = invert_h(h,g_len); %invert h 
+    %convolve g with the trace of the whole day and use it to calculate an
+    %adaptive threshold for detecting events
+    trace_filtered = conv(trace,g,'same');
+    idx_event_frames_out = extract_candidate_events(trace_filtered,fine_thresh_multiplier,0);
+    out = {idx_event_frames_out,h,g};
+end
+
+function event_mat = construct_events(idx_event_frames)
+
+%align all the trial segments(where there is an event) and accumulate all of them.
+%The events in event_mat are of fixed length with frames before the maximum and
+%frames after the maximum determind by the function inputs.
```

Comment:
Be clearer with regards to the format of `event_mat`. Is it a matrix where each row represents an event waveform?